### DISCUSSION TITLE
Usage with storybook

### DISCUSSION THREAD

**Discussion by tim-richter** [vanilla-extract-css/vanilla-extract#371] [Ideas]
Hey everyone :wave: 

I've made storybook work with vanilla-extract and wanted to introduce the idea of adding documentation about the process so other developers don't run into the same issues I had with it.

Problems were mainly:
- The mini-css-extract-plugin relies on webpack 5 which storybook currently only supports by enabling it manually ([here](https://gist.github.com/shilman/8856ea1786dcd247139b47b270912324#upgrade) is a guide on how to do it)
- When extending the default storybook webpack config the .css rule that storybook comes with casts a pretty wide net and matches the vanilla.css files which results in errors because css-loader is applied multiple times to those files. My solution involved just filtering out the .css rules from the storybook config and then adding the vanilla.css rule (given the code for this seems rather brittle and would probably need to be adjusted) [Line 18](https://github.com/tim-richter/next-boiler/blob/main/.storybook/vanilla-extract.js#L18)

As far as I can see everything works perfectly fine now even hot reloading etc.

[Reproduction Repo](https://github.com/tim-richter/next-boiler/tree/main/.storybook)

**Comment by tmm** [comment]
Also worth noting that Storybook has [experimental Vite support](https://storybook.js.org/blog/storybook-for-vite/). I tried it out and it works well — and is super fast.

**Reply by w3b-wiz4rd** [reply]
@tmm Heya, by any change do you have an example of how to set it up  ? I've been struggling to make it work.
My current setup is: Vite + Storybook  + vanilla-extract

If I use the component on the App.tsx generated by Vite everything works.

<img width="1839" alt="Screen Shot 2022-06-22 at 14 15 34" src="https://user-images.githubusercontent.com/17417706/175118373-54e35acd-2fac-41ae-a87f-e34e085f405d.png">


But in storybook I keep getting this:

<img width="1910" alt="Screen Shot 2022-06-22 at 14 15 44" src="https://user-images.githubusercontent.com/17417706/175118384-0c31d92d-fa12-46c0-8194-5cc1358559f6.png">


Is the same component .. but still not working on storybook.


I followed the same pattern, but removing the miniCSS plugin, don't think is needed on vite;

This is my current man.js

`const path = require("path")

module.exports = {
  stories: ["../src/**/*.stories.mdx", "../src/**/*.stories.@(js|jsx|ts|tsx)"],
  addons: [
    path.resolve("./.storybook/vanilla-extract"),
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@storybook/addon-interactions",
  ],

  framework: "@storybook/react",
  core: {
    builder: "@storybook/builder-vite",
  },
  features: {
    storyStoreV7: true,
  },
}
`

This is my current vanilla-extract.js

`const { vanillaExtractPlugin } = require("@vanilla-extract/vite-plugin")

module.exports = {
  webpackFinal(baseConfig, options) {
    const { module = {}, plugins = {} } = baseConfig

    return {
      ...baseConfig,
      plugins: [...plugins, new vanillaExtractPlugin()],
    }
  },
}
`
Any help would be appreciate it.
Regards.

**Reply by fredericoo** [reply]
> @tmm Heya, by any change do you have an example of how to set it up ? I've been struggling to make it work. My current setup is: Vite + Storybook + vanilla-extract
> 
> If I use the component on the App.tsx generated by Vite everything works.
> 
> <img alt="Screen Shot 2022-06-22 at 14 15 34" width="1839" src="https://user-images.githubusercontent.com/17417706/175118373-54e35acd-2fac-41ae-a87f-e34e085f405d.png">
> 
> But in storybook I keep getting this:
> 
> <img alt="Screen Shot 2022-06-22 at 14 15 44" width="1910" src="https://user-images.githubusercontent.com/17417706/175118384-0c31d92d-fa12-46c0-8194-5cc1358559f6.png">
> 
> Is the same component .. but still not working on storybook.
> 
> I followed the same pattern, but removing the miniCSS plugin, don't think is needed on vite;
> 
> This is my current man.js
> 
> `const path = require("path")
> 
> module.exports = { stories: ["../src/**/*.stories.mdx", "../src/**/*.stories.@(js|jsx|ts|tsx)"], addons: [ path.resolve("./.storybook/vanilla-extract"), "@storybook/addon-links", "@storybook/addon-essentials", "@storybook/addon-interactions", ],
> 
> framework: "@storybook/react", core: { builder: "@storybook/builder-vite", }, features: { storyStoreV7: true, }, } `
> 
> This is my current vanilla-extract.js
> 
> `const { vanillaExtractPlugin } = require("@vanilla-extract/vite-plugin")
> 
> module.exports = { webpackFinal(baseConfig, options) { const { module = {}, plugins = {} } = baseConfig
> 
> ```
> return {
>   ...baseConfig,
>   plugins: [...plugins, new vanillaExtractPlugin()],
> }
> ```
> 
> }, } ` Any help would be appreciate it. Regards.

Did you find a solution?

**Reply by neolectron** [reply]
[Solution here](https://github.com/vanilla-extract-css/vanilla-extract/discussions/371#discussioncomment-3676198)

**Comment by SaschaDens** [comment]
Thanks @tim-richter for the reproduction repository!

I was exactly trying to do the same things but then on the current release of storybook (with Webpack 4) but got eventually stuck.
I'm wondering if there is a possibility to make `vanilla-extract` run without the `mini-css-extract-plugin` loader where the `style-loader`is used instead just to make the components visible in storybook.

No luck so far 😞 but I'll continue to also try as this looks promising.

**Comment by SaschaDens** [comment]
Good news I was able to make `vanilla-extract` run in the current release of storybook with the following adaptations.
I'll post it here for other developers in a similar situation 😄 .

For `mini-css-extract-plugin@2.3.0` you'll need Webpack 5 for the current release but I've tested with `mini-css-extract-plugin@1.6.2` and this also seems to do the trick.

```bash
yarn add --dev @vanilla-extract/css @vanilla-extract/webpack-plugin mini-css-extract-plugin@1.6.2
```

Then similar as @tim-richter mentioned the storybook configuration needs to be extended to exclude `.vanilla.css` out of the `/\.css$/` regex of storybook and include the `@vanilla-extract/webpack-plugin` plugin.

This can be done by adapting storybooks following files:
```javascript
// ./.storybook/main.js
const path = require('path');

module.exports = {
  stories: ['../src/**/*.stories.mdx', '../src/**/*.stories.@(js|jsx|ts|tsx)'],
  addons: [
    path.resolve('./.storybook/vanilla-extract.js'),
    '@storybook/addon-links',
    '@storybook/addon-essentials',
    '@storybook/preset-create-react-app',
  ],
};

```
Then in `./.storybook/vanilla-extract.js` you inject the plugin
```javascript
// ./.storybook/vanilla-extract.js
const { VanillaExtractPlugin } = require('@vanilla-extract/webpack-plugin');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');

module.exports = {
  webpackFinal(baseConfig, options) {
    const { module = {}, plugins = {} } = baseConfig;

    const isCssRule = (rule) =>
      []
        .concat(rule?.test)
        .filter(Boolean)
        .find((expr) => expr.test('random.css'));
    module.rules
      .map(({ oneOf, ...rule }) =>
        oneOf ? oneOf.find(isCssRule) : isCssRule(rule),
      )
      .filter(Boolean)
      // Here we exclude *.vanilla.css files which will be taken up by the vanilla-extract
      .forEach((match) => match.exclude.push(/\.vanilla\.css$/));

    return {
      ...baseConfig,
      plugins: [
        ...plugins,
        new VanillaExtractPlugin(),
        new MiniCssExtractPlugin(),
      ],
      module: {
        ...module,
        rules: [
          ...module.rules,
          {
            test: /\.vanilla\.css$/i, // Targets only CSS files generated by vanilla-extract
            use: [
              MiniCssExtractPlugin.loader,
              {
                loader: require.resolve('css-loader'),
                options: {
                  url: false, // Required as image imports should be handled via JS/TS import statements
                },
              },
            ],
          },
        ],
      },
    };
  },
};
```
### For storybook with webpack@5

@tim-richter for [Line 18](https://github.com/tim-richter/next-boiler/blob/862dda907f20123e587a5f33a0a2f2a75cf68e77/.storybook/vanilla-extract.js#L18) you disable the default `css-loader`. You could potentially use the following snippet instead to make the regex ignore also `\.vanilla.css$` and make both options work.
```javascript
const { VanillaExtractPlugin } = require('@vanilla-extract/webpack-plugin');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');

module.exports = {
  webpackFinal(baseConfig, options) {
    const { module = {}, plugins = {} } = baseConfig;

    const cssRule = module.rules.find((rule) => rule?.test?.test('test.css'));
    cssRule.test = /.*(?<!\.vanilla)\.css$/;

    return {
      ...baseConfig,
      plugins: [
        ...plugins,
        new VanillaExtractPlugin(),
        new MiniCssExtractPlugin(),
      ],
      module: {
        ...module,
        rules: [
          ...module.rules,
          {
            test: /\.vanilla\.css$/i, // Targets only CSS files generated by vanilla-extract
            use: [
              MiniCssExtractPlugin.loader,
              {
                loader: require.resolve('css-loader'),
                options: {
                  url: false, // Required as image imports should be handled via JS/TS import statements
                },
              },
            ],
          },
        ],
      },
    };
  },
};
```

**Reply by tim-richter** [reply]
Really cool! Will definitely try that out seems like a better approach with the regex 😉

**Reply by llm-gourmet** [reply]
FUIYOOOH that worked out of the noodle box!

**Reply by cadbox1** [reply]
I ran into the same issue where removing the original css loader meant regular css files couldn't load.

I tried this snippet but it still didn't work so I ended up modifying the added css-loader to include all css files and it seemed to work. maybe something changed on storybook's side.

https://github.com/tim-richter/next-boiler/pull/27

**Comment by sean-tmg** [comment]
@tim-richter thanks so much for putting this together. Huge time saver for me to pick out the bits.

I used @SaschaDens suggestion for Webpack 5, and also had to add this to my Storybook [`main.js`](https://github.com/storybookjs/storybook/issues/15336#issuecomment-888528747)

♥️

**Comment by pruthvi2103** [comment]
For anyone struggling to get storybook up and running I have a simple example repo which has Storybook and Vanilla Extact with Vite. Feel free to go through the [repo](https://github.com/pruthvi2103/vanilla-extract-with-storybook). Its simple, hassle free and works well.

**Reply by w3b-wiz4rd** [reply]
@pruthvi2103 Thanks a lot man! I've been struggling with this setup! It works!

**Comment by wallacerenan** [comment]
Best solution with webpack4!

create a file called webpack.config.js inside the .storybook folder, with the content:

```js
// .storybook/webpack.config.js
const { VanillaExtractPlugin } = require('@vanilla-extract/webpack-plugin');

module.exports = ({ config }) => {

  const plugins = [new VanillaExtractPlugin(), ...config.plugins]

  config.plugins = plugins

  return config;
};
```

If you need another plugin, example:

```js
// .storybook/webpack.config.js
const { VanillaExtractPlugin } = require('@vanilla-extract/webpack-plugin');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');

module.exports = ({ config }) => {

  const plugins = [new VanillaExtractPlugin(), new MiniCssExtractPlugin(), ...config.plugins]

  config.plugins = plugins

  return config;
};
```

**Comment by liam-jones-lucout** [comment]
For anyone struggling to get Vanilla-Extract working with Storybook Vite, here's my storybook main.js:
```javascript
const { mergeConfig } = require('vite');

module.exports = {
  "stories": [
    "../src/**/*.stories.mdx",
    "../src/**/*.stories.@(js|jsx|ts|tsx)"
  ],
  "addons": [
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@storybook/addon-interactions"
  ],
  "framework": "@storybook/react",
  "core": {
    "builder": "@storybook/builder-vite"
  },
  "features": {
    "storyStoreV7": true
  },
  async viteFinal(config) {
    return mergeConfig(config, {
      plugins: [require("@vanilla-extract/vite-plugin").vanillaExtractPlugin()]
    });
  }
}
```

**Comment by jd-oconnor** [comment]
I wanted to share this in case anyone runs into a similar issue with SB 7 and getting build errors.

I spent several hours over the past couple of days trying to get SB 7 to build. I've been working on a private package for my company and implementing SB and Chromatic. This thread helped with some things, but once I got this error, I hit a brick wall
![Screen_Shot_2022-12-13_at_5 17 15_PM](https://user-images.githubusercontent.com/44309632/207631603-99c24df9-1261-4a4f-b4bf-f297c7606a8c.png)

Running it locally was completely fine. But every time I ran a build, that's the error I would get.

Turns out, VE ships an older version  of `esbuild` (0.11.16) in `@vanilla-extract/integration` but SB 7 needs/uses a newer version and that's what's causing the issue with SB. Kudos to the SB maintainer I worked with that helped me figure it out. Simple fix was adding:

```json
"overrides": {
  "esbuild": "0.16.6"
},
```

Running it locally still works and the build now succeeds!

Here's my `main.js`
```js
const path = require('path');

module.exports = {
  "core": {
    "builder": "webpack5",
  },
  "stories": [
    "../src/**/*.stories.mdx",
    "../src/**/*.stories.@(js|jsx|ts|tsx)"
  ],
  "addons": [
    path.resolve('./.storybook/vanilla-extract.js'),
    path.resolve('./.storybook/ts-paths'),
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@storybook/addon-interactions"],
  "framework": {
    name: "@storybook/react-webpack5",
    options: {}
  },
  docs: {
    docsPage: "automatic"
  },
  typescript: {
    check: false,
    checkOptions: {},
    reactDocgen: false,
    reactDocgenTypescriptOptions: {
       shouldExtractLiteralValuesFromEnum: true,
       propFilter: (prop) => (prop.parent ? !/node_modules/.test(prop.parent.fileName) : true),
    },
  },
};
```

**Comment by kim-jaedeok** [comment]
I think it could be simpler.
here is my main.js

```js
const { VanillaExtractPlugin } = require("@vanilla-extract/webpack-plugin");
const { merge } = require("webpack-merge");

module.exports = {
  stories: ["../src/**/*.stories.mdx", "../src/**/*.stories.@(js|jsx|ts|tsx)"],
  addons: [
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@storybook/addon-interactions",
  ],
  framework: "@storybook/react",
  core: {
    builder: "@storybook/builder-webpack5",
  },
  webpackFinal: async (config) =>
    merge(config, {
      plugins: [new VanillaExtractPlugin()],
    }),
};

```

**Comment by dachewster99** [comment]
It takes roughly 15 seconds between making code changes, and actually seeing those changes in Storybook. Is anybody else having these issues? I do have a large amount of stories, but 15 seconds to notice a change still seems pretty excessive. 

I dont believe this to be storybook itself, as if I update a prop in the side panel, it updates immediately. So i am assuming its something in the build. I am using vite

**Reply by jd-oconnor** [reply]
Hmmm 🤔 definitely wouldn't expect to have that with vite. Is there a repo you can share to take a look at the code?

**Reply by yyyyaaa** [reply]
same here, it gets worse once you get past 30+ stories (I'm using vite + storybook)

**Comment by cseas** [comment]
Thank you! Your solution helped point me in the right direction. However, instead of removing Storybook's css rule, I found it better to just add an `exclude` property to it instead so it leaves vanilla extract's files alone.

Here's my working config with Storybook 7 and webpack 5:

```js
// .storybook/main.js
import { VanillaExtractPlugin } from "@vanilla-extract/webpack-plugin";
import MiniCssExtractPlugin from "mini-css-extract-plugin";

export default {
  // ... other options
  framework: {
    name: "@storybook/react-webpack5",
    options: {},
  },
  webpackFinal: async (config) => {
    const customConfig = { ...config };

    // modify css rule of storybook to exclude Vanilla Extract files
    for (let i = 0; i < customConfig.module.rules.length; i++) {
      if (customConfig.module.rules[i].test.test("sample.css")) {
        customConfig.module.rules[i].exclude = /\.vanilla\.css$/i;
        break;
      }
    }

    customConfig.module.rules.push(
      {
        test: /\.vanilla\.css$/i,
        use: [
          MiniCssExtractPlugin.loader,
          {  loader: "css-loader", options: {  url: false  }  },
        ],
      },
    );

    customConfig.plugins.push(
      new VanillaExtractPlugin(),
      new MiniCssExtractPlugin()
    );

    return customConfig;
  },
};
```
Full config for reference is [here](https://github.com/hazel-ui/hazel-ui/blob/main/.storybook/main.js).

**Comment by mszczepanczyk** [comment]
Storybook 6 + Webpack 5 with working HMR (#905)

```js
// .storybook/main.js

module.exports = {
  ...
  webpackFinal: async (config) => ({
    ...config,
    optimization: {
      ...config.optimization,
      // https://github.com/vanilla-extract-css/vanilla-extract/issues/905
      splitChunks: false,
    },
    plugins: [...config.plugins, new VanillaExtractPlugin()],
  })
}
```

**Reply by airtonix** [reply]
doesn't work on storybook 7

edit: `it does, but...`

went with: 

```ts
import type { StorybookConfig } from '@storybook/nextjs';
import { VanillaExtractPlugin } from '@vanilla-extract/webpack-plugin';
import MiniCssExtractPlugin from 'mini-css-extract-plugin';

const config: StorybookConfig = {
  framework: {
    name: '@storybook/nextjs',
    options: {},
  },
  core: {
    builder: '@storybook/builder-webpack5',
  },
  stories: [
    '../../website/src/**/*.stories.@(js|jsx|ts|tsx|mdx)',
    '../../../pkg/ds/*/src/lib/**/*.stories.@(js|jsx|ts|tsx|mdx)',
  ],
  addons: ['@storybook/addon-essentials', '@nrwl/react/plugins/storybook'],
  webpackFinal: async (config) => {
    return {
      ...config,
      optimization: {
        ...config.optimization,
        // https://github.com/vanilla-extract-css/vanilla-extract/issues/905
        splitChunks: false,
      },
      plugins: [
        ...config.plugins,
        new VanillaExtractPlugin(),
        new MiniCssExtractPlugin(),
      ],
    };
  },
};

export default config;
```

using nxdev here.

**Comment by gabrieloureiro** [comment]
This Vite solutions doesn't work for me. Any update?

![Screen Shot 2023-03-07 at 10 43 38](https://user-images.githubusercontent.com/39869298/223440124-7bf822ec-92aa-4aac-9c27-2066b8969a37.png)
![Screen Shot 2023-03-07 at 10 44 07](https://user-images.githubusercontent.com/39869298/223440226-fadfe037-087a-4345-a206-774890d4cedf.png)

**Comment by domyen** [comment]
@gabrieloureiro check out [this recipe](https://storybook.js.org/recipes/@vanilla-extract/css) from the Storybook team. 

Since Storybook reads your vite config out of the box, try removing `viteFinal`

> For Vite users, Storybook uses your project's vite.config.js for its build configuration. This means if your project uses Vanilla Extract, it's also already configured for your Storybook. 🎉

> You can view a full example on [Stackblitz](https://stackblitz.com/edit/github-cytqag?file=.storybook/main.ts).
